### DISCUSSION TITLE
Regex Continued

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#4139] [Development]
**Task**: Wrap up the remainder of the regex uses

- Create `struct Regex` rather than using `regcomp()` or `REG_COMP()`
- Use `mutt_regex_match()` or `mutt_regex_capture()` instead of `regexec()`

We'll need to expose a few `regcomp()`/`regexec()` flags (as `DT_REGEX_*`)

- `REG_ICASE` - Ignore case
- `REG_NEWLINE` - Don't match newline
- `REG_NOTBOL` - Don't match beginning-of-line

## Easy

These uses are limited to a single function, or `case` statement.

### tls_check_stored_hostname()

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/conn/ssl_gnutls.c#L218

### mutt_buffer_select_file()

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/browser.c#L2100-L2101
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/browser.c#L2123

### mutt_yesorno()

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/curs_lib.c#L412
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/curs_lib.c#L417

## Medium

These involve upgrading a struct member to use `struct Regex`

### struct ColorLine

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/index.c#L869
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/pager.c#L1032
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/pager.c#L1129-L1131
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/pager.c#L1207-L1208

### struct AttachMatch

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/mutt_parse.c#L96-L97

### struct RangeRegex

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/pattern.c#L945

### struct Pattern

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/pattern.c#L1072

## Hard

These are hard because they span quite a lot of code

### struct PagerRedrawData

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/pager.c#L1756-L1757

### struct Menu

https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/browser.c#L918
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/browser.c#L923
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/menu.c#L960
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/query.c#L237
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/query.c#L239
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/query.c#L244
https://github.com/neomutt/neomutt/blob/5cd5560b8f4348a7a19ea5afbab59775e400f0e2/query.c#L248

**Comment by psyomn** [comment]
So I'd like to start working on this. Can I make multiple PRs to address these changes piecemeal?

**Comment by flatcap** [comment]
> Can I make multiple PRs to address these changes piecemeal?

Yes, if you want.

Also, if you create branches on this repo, named 'devel/SHORT-NAME`, that makes it a little easier for gahr and I to review/tidy/merge.

Keep asking questions.  Thanks.
